### PR TITLE
Initial checkin of Jaeger adapter

### LINKED ISSUE #7317: Support JaegerTracer
Prototype different implementation approaches on how to integrate Jaeger into OpenLiberty.

### PR DESCRIPTION
**Pull request by fmhwong**
Fixes #7317

Using the adapter approach, this avoided direct reference of Jaeger client library classes.  The Jaeger jars are not required to be shipped in OpenLiberty.  They should be included as shared libraries in server.xml

### CODE REVIEW

**Comment by fmhwong** [comment]
#build

**Comment by fmhwong** [comment]
#build

**Comment by tevans78** [comment]
@fmhwong You should consider delivering this initially as a noship or beta auto feature until it is 100% ready for GA.

**Comment by fmhwong** [comment]
Converted to beta feature `jaeger-1.0` 
#build

File: dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+JAEGER_PASSWORD_CANNOT_DECODE=CWMOT1003W: The value of JAEGER_PASSWORD cannot be decoded.
+JAEGER_PASSWORD_CANNOT_DECODE.explanation=The password cannot be decoded and therefore it is ignored.
+JAEGER_PASSWORD_CANNOT_DECODE.useraction=Use securityUtility to encode passwords.
+
+JAEGER_TAGS_CANNOT_PARSE=CWMOT1004W: The value of JAEGER_TAGS cannot be parsed.
+JAEGER_TAGS_CANNOT_PARSE.explanation=The tags cannot be parsed and therefore they are ignored.
+JAEGER_TAGS_CANNOT_PARSE.useraction=Use a comma separated name=value pairs.
+
+JAEGER_CONFIG_EXCEPTION=CWMOT1005E: JaegerTracer cannot be instantiated.  Reason: {0} 
+JAEGER_CONFIG_EXCEPTION.explanation=JaegerTracer cannot be instantiated due to configuration issues.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops]: Change
"JaegerTracer cannot be instantiated due to configuration issues."
To
"A JaegerTracer object cannot be instantiated due to configuration issues."

File: dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#-----------------------------------------------------------------------------------------------------------------------------
+JAEGER_TRACER_CREATED=CWMOT1001I: An instance of JaegerTracer has been created for application {0}.  Tracing information is sent to {1}.
+JAEGER_TRACER_CREATED.explanation=An instance of JaegerTracer has been created to send spans from the application to the tracing backend.
+JAEGER_TRACER_CREATED.useraction=No action required.
+
+JAEGER_ENV_VAR_PARSE_ERROR=CWMOT1002W: The value of system property or environment variable {0} is not a valid type of {1}.
+JAEGER_ENV_VAR_PARSE_ERROR.explanation=The value defined in the system property or environment variable cannot be converted to the type.
+JAEGER_ENV_VAR_PARSE_ERROR.useraction=Define a value for the system property or environment variable that matches the type. 
+
+JAEGER_PASSWORD_CANNOT_DECODE=CWMOT1003W: The value of JAEGER_PASSWORD cannot be decoded.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops]: Change
"The value of JAEGER_PASSWORD cannot be decoded."
To
"The value of the JAEGER_PASSWORD environment variable cannot be decoded."

File: dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+JAEGER_PASSWORD_CANNOT_DECODE.explanation=The password cannot be decoded and therefore it is ignored.
+JAEGER_PASSWORD_CANNOT_DECODE.useraction=Use securityUtility to encode passwords.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.microprofile.opentracing.jaeger.adapter/resources/com.ibm.ws.microprofile.opentracing.jaeger.resources/JaegerMessages.nlsprops]: Change
"Use securityUtility to encode passwords."
To
"Use the securityUtility command to encode passwords."

**Comment by Azquelt** [comment]
You appear to have put all of the code in the same bundle?

I'm not a huge fan of that. It would be easier to make sure you don't accidentally import something that you shouldn't if you separate the code that is allowed to directly access Jaeger from the code which is not.

I _think_ you also run the risk of weird class cast exceptions if you were to introduce a new interface to the adaptor but forget to add it to the list of interfaces given to the AppLibraryClassLoader.

**Comment by Azquelt** [comment]
You've mentioned in the text for this PR that you expect Jaeger to be provided as a shared library in the server.xml.

However, the approach you've taken merely requires Jaeger to be available on the application class path. I.e. it could be in a library linked to the application in the server.xml, but it could also be packaged in the application itself.

Is this the desired behaviour?

**Comment by Azquelt** [comment]
Where are the FAT tests for this functionality?

File: dev/com.ibm.ws.opentracing.1.3/src/com/ibm/ws/opentracing/OpentracingTracerManager.java [hunk 4]
```diff
@@ -72,15 +72,17 @@ private static Tracer ensureTracer(String appName) {
...
                 tracerCase = "newly created";
             } else {
                 tracerCase = "previously created";
             }
         }
 
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
+
+        {
```
Comment (Azquelt) [inline dev/com.ibm.ws.opentracing.1.3/src/com/ibm/ws/opentracing/OpentracingTracerManager.java]: This whitespace change looks odd

Comment (fmhwong) [reply dev/com.ibm.ws.opentracing.1.3/src/com/ibm/ws/opentracing/OpentracingTracerManager.java]: Cleaned up the whitespace.

**Comment by fmhwong** [comment]
#build

File: dev/com.ibm.ws.microprofile.opentracing.jaeger/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  com.ibm.ws.microprofile.opentracing.jaeger;version=1.0
+
+Import-Package: \
+  *
+
+Private-Package: com.ibm.ws.microprofile.opentracing.jaeger.resources
+
+Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
+
+instrument.disabled: true
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.opentracing.jaeger/bnd.bnd]: I don't think this should be necessary any more.

**Comment by fmhwong** [comment]
The failure in PB is in JSONB.  It looks like it has been reported in defect 263359.

**Pull request merged**